assets/src/edit-story/elements/text/edit.js

```diff
@@ -0,0 +1,147 @@
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import { Editor, EditorState, SelectionState } from 'draft-js';
+import { stateFromHTML } from 'draft-js-import-html';
+import { stateToHTML } from 'draft-js-export-html';
+
+/**
+ * WordPress dependencies
+ */
+import { useState, useEffect, useLayoutEffect, useRef, useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useStory } from '../../app';
+import { useCanvas } from '../../components/canvas';
+import {
+	ElementWithPosition,
+	ElementWithSize,
+	ElementWithFont,
+	ElementWithBackgroundColor,
+	ElementWithFontColor,
+} from '../shared';
+import { getFilteredState, getHandleKeyCommand } from './util';
+
+const Element = styled.div`
+	margin: 0;
+	${ ElementWithPosition }
+	${ ElementWithSize }
+	${ ElementWithFont }
+	${ ElementWithBackgroundColor }
+	${ ElementWithFontColor }
+
+	&::after {
+		content: '';
+		display: block;
+		position: absolute;
+		left: 0;
+		top: 0;
+		width: 100%;
+		height: 100%;
+		border: 1px solid ${ ( { theme } ) => theme.colors.mg.v1 }70;
+		pointer-events: none;
+	}
+`;
+
+function TextEdit( { content, color, backgroundColor, width, height, x, y, fontFamily, fontSize, fontWeight, fontStyle } ) {
+	const props = {
+		color,
+		backgroundColor,
+		fontFamily,
+		fontStyle,
+		fontSize,
+		fontWeight,
+		width,
+		height,
+		x,
+		y,
+	};
+	const { actions: { setPropertiesOnSelectedElements } } = useStory();
+	const { state: { editingElementState } } = useCanvas();
+	const { offset, clearContent } = editingElementState || {};
+	// To clear content, we can't just use createEmpty() or even pure white-space.
+	// The editor needs some content to insert the first character in,
+	// so we use a non-breaking space instead and trim it on save if still present.
+	const EMPTY_VALUE = '\u00A0';
+	const initialState = (
+		clearContent ?
+			EditorState.createWithContent( stateFromHTML( EMPTY_VALUE ) ) :
+			EditorState.createWithContent( stateFromHTML( content ) )
+	);
+	const [ editorState, setEditorState ] = useState( initialState );
+	const mustAddOffset = useRef( offset ? 2 : 0 );
+
+	// This is to allow the finalizing useEffect to *not* depend on editorState,
+	// as would otherwise be a lint error.
+	const lastKnownState = useRef( null );
+
+	// This filters out illegal content on paste and updates state accordingly.
```

Comment:
What does illegal content mean in this context?

Reply:
[Added some context](https://github.com/ampproject/amp-wp/pull/3804/files#diff-5f4f624867fd96363e2603afb736f301R85-R86).

Reply:
[Added some context](https://github.com/ampproject/amp-wp/pull/3804/files#diff-5f4f624867fd96363e2603afb736f301R85-R86).